cpp/oneapi/dal/test/engine/linalg/dot.hpp

```diff
@@ -0,0 +1,50 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/test/engine/linalg/matrix.hpp"
+
+namespace oneapi::dal::test::engine::linalg {
+
+template <typename Float, layout lyt_a, layout lyt_b, layout lyt_c>
+struct dot_op {
+    void operator()(const matrix<Float, lyt_a>& a,
+                    const matrix<Float, lyt_b>& b,
+                    matrix<Float, lyt_c>& c,
+                    Float alpha,
+                    Float beta) const;
+};
+
+template <typename Float, layout lyt_a, layout lyt_b, layout lyt_c>
+inline void dot(const matrix<Float, lyt_a>& a,
+                const matrix<Float, lyt_b>& b,
+                matrix<Float, lyt_c>& c,
+                Float alpha = Float(1),
+                Float beta = Float(0)) {
+    dot_op<Float, lyt_a, lyt_b, lyt_c>{}(a, b, c, alpha, beta);
+}
+
+template <typename Float, layout lyt_a, layout lyt_b, layout lyt_c = layout::row_major>
```

Comment:
What if user wants to change layout of `c` - does he required to specify all the template parameters before `lyt_c`?

Reply:
Obviously yes

Reply:
If "special" layout for `c` is needed, there is `dot` overload that accepts `c` by reference, so template parameters can be deduced.